### DISCUSSION TITLE
Setting Pin value of DigitalOutput programmatically

### DISCUSSION THREAD

**Discussion by kevinbolding** [bonsai-rx/bonsai#1240] [General]
This question is copied over from the Discord:

Is there a way to programmatically control the Pin value for a DigitalOutput? The value could be generated any number of ways (by a random integer generator or by reading in from a list in a separate protocol file). 

To make this concrete, we would like to control multiple solenoids in an odor delivery system with different Arduino pins. We have gotten to the stage of doing this one by one in Bonsai (using Firmata and DigitalOutput), but for real protocols/scripts we'll need to select a combination of pins to flip to High on each trial. Any advice appreciated.

**Comment by bruno-f-cruz** [comment]
I don't have an Arduino to test, but I think you can create an observable for each pin and filter an incoming tuple<int, bool> that specifies which pin you are targeting. Check the attached workflow for a barebones example of what this would look like.
[InstantiatePins.zip](https://github.com/bonsai-rx/bonsai/files/10576709/InstantiatePins.zip)
